### PR TITLE
Enhancements

### PR DESCRIPTION
**Pull request by fmido88**
- Fix instate of adding new override, editing the old one, so prevent the error of multiple records exists when calling $DB->get_record().
- Adding fields to the database table for time updated and overrider (the person did the override process).
- Adding additional method for overriding by adding additional maxviews instated of complete override.
- Adding capability for overriding, so only users with this capability can perform the override.
- Adding Events for overriding so that manager can track any override process from the log.

I'm looking forward for your reply after reviewing these changes.

### CODE REVIEW

**Comment by fmido88** [comment]
Hello @danielneis
It will be great and means alot for me if you take a look for these changes.

**Comment by danielneis** [comment]
Hello, @fmido88 
I'll do my best to review it by the next week.

**Comment by fmido88** [comment]
Thanks @danielneis
I appreciate it very much and waiting to tell me if there is any modifications you need.

File: classes/form/override.php [hunk 1]
```diff
@@ -40,33 +44,81 @@ public function definition() {
 
         $modinfo = get_fast_modinfo($this->_customdata['courseid']);
         $options = ['' => ''];
+
         foreach ($modinfo->cms as $i) {
+            // Filter course modules by that only has maxviews conditions.
+            $getviewslimit = $DB->get_field_select('course_modules', 'availability' ,
```
Comment (danielneis) [inline classes/form/override.php:50]: For sure there is availability information already on $i , it would be nice not to have 1 more query per course module. could please check and change to use the appropriate attribute/function ?

File: lang/en/availability_maxviews.php [hunk 2]
```diff
@@ -22,24 +22,47 @@
...
 $string['ajaxerror'] = 'Error contacting server';
 $string['confirmdeleteoverride'] = 'Are you sure you want to delete the override for {$a}? This action cannot be undone.';
 $string['coursemodule'] = 'Course module';
+$string['coursemodule_help'] = 'select the course module needed to override it.';
 $string['description'] = 'Prevent access after user view the module a certain number of times.';
 $string['editingoverride'] = 'Editing override';
 $string['eithernotdescription'] = 'you have not reached the limit of {$a->viewslimit} views (you have {$a->viewscount} views)';
 $string['eitherdescription'] = 'you have reached the limit of {$a->viewslimit} views (you have {$a->viewscount} views)';
+$string['event_maxviews_override_created'] = 'The limitation of maxviews override created';
+$string['event_maxviews_override_description'] = 'The maxviews of user of id {$a->relateduserid} in course module of id {$a->cmid}
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:35]: please, change the string to read "user with id" , "course with id"

File: lang/en/availability_maxviews.php [hunk 3]
```diff
@@ -22,24 +22,47 @@
...
+override {$a->type} by user of id {$a->userid} and {$a->reset}';
+$string['event_maxviews_override_updated'] = 'The limitation of maxviews override updated';
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:37]: please, remove the initial "The"

File: lang/en/availability_maxviews.php [hunk 4]
```diff
@@ -22,24 +22,47 @@
...
+$string['event_maxviews_override_deleted'] = 'The limitation of maxviews override deleted';
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:38]: please, remove the initial "The"

File: lang/en/availability_maxviews.php [hunk 5]
```diff
@@ -22,24 +22,47 @@
...
+$string['overridetype_desc'] = 'Normal override: discarding the old condition and set a new one for the selected users.
+
+Adding maxviews: This mean that you add or subtract the new value from the old maxviews.';
 $string['overrides'] = 'Overrides';
 $string['overrideupdated'] = 'Override updated';
 $string['participant'] = 'Participant';
+$string['participant_help'] = 'Select the participants needed to override their maximum views for the selected course module.';
 $string['pluginname'] = 'Max Views';
 $string['resetviews'] = 'Reset views';
+$string['resetviews_help'] = 'If this options checked, the views count for the selected users will reseted to zero for the selected course module.';
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:63]: it should be "will be reset to zero"

File: lang/en/availability_maxviews.php [hunk 6]
```diff
@@ -22,24 +22,47 @@
...
+$string['resetyes'] = 'maxviews has been rest';
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:64]: we have a typo here, it should be "has been reset"

File: lang/en/availability_maxviews.php [hunk 7]
```diff
@@ -22,24 +22,47 @@
...
+$string['resetno'] = 'maxviews not reset';
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:65]: it should be "hasn't been reset"

File: classes/form/override.php [hunk 8]
```diff
@@ -40,33 +44,81 @@ public function definition() {
...
-        $mform->setType('maxviews', PARAM_TEXT);
+        if ($type === 'normal') { // Normal override.
+            $mform->addElement('text', 'maxviews', get_string('maxviews', 'availability_maxviews'));
+            $mform->setType('maxviews', PARAM_INT);
+            $mform->addHelpButton('maxviews', 'maxviews', 'availability_maxviews');
+        } else { // Additional maxviews.
+            $mform->addElement('text', 'maxviews', get_string('maxviews_add', 'availability_maxviews'));
+            $mform->setType('maxviews', PARAM_INT);
+            $mform->addHelpButton('maxviews', 'maxviews_add', 'availability_maxviews');
+        }
```
Comment (danielneis) [inline classes/form/override.php:121]: this could be simpler if you move the calling of addElement, setType and addHelpButton outside of the if condition, and inside you create variables that will have the values used in the function calls after the if/else.
like
 $label = get_string('max_views' ...
$helpstrkey = 'max_views' (or 'max_views_add')
and the setType is always the same

File: classes/output/index.php [hunk 9]
```diff
@@ -60,11 +64,19 @@ public function export_for_template(renderer_base $output) {
...
+            $overrides[$key]->date = max($o->timecreated, $o->timeupdated);
         }
+
         return (object)[
             'overrides' => array_values($overrides),
+            'typeadd' => $typeadd,
             'hasoverrides' => !empty($overrides),
-            'newoverrideurl' => (new moodle_url('/availability/condition/maxviews/override.php', ['courseid' => $this->courseid]))->out(false),
+            'newoverrideurl' => (new moodle_url('/availability/condition/maxviews/override.php',
+                                                    ['courseid' => $this->courseid]))->out(false),
```
Comment (danielneis) [inline classes/output/index.php:79]: please, indent correctly

File: db/upgrade.php [hunk 10]
```diff
@@ -57,5 +54,36 @@ function xmldb_availability_maxviews_upgrade($oldversion) {
...
+        // Define the new field for timecreated.
+        $field = new xmldb_field('timecreated', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'maxviews');
+        // Conditionally lunch add field.
+        if (!$dbman->field_exists($table, $field)) {
+            $dbman->add_field($table, $field);
+        }
+
+        // Define the new field for timemodified.
+        $field = new xmldb_field('timeupdated', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'timecreated');
+        // Conditionally lunch add field.
```
Comment (danielneis) [inline db/upgrade.php:79]: "launch" :)

File: db/upgrade.php [hunk 11]
```diff
@@ -57,5 +54,36 @@ function xmldb_availability_maxviews_upgrade($oldversion) {
...
+        $field->setComment('The id of the user who did the overriding process.');
+
+        // Conditionally lunch add field.
+        if (!$dbman->field_exists($table, $field)) {
+            $dbman->add_field($table, $field);
+        }
+
+        // Define the new field for timecreated.
+        $field = new xmldb_field('timecreated', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'maxviews');
+        // Conditionally lunch add field.
```
Comment (danielneis) [inline db/upgrade.php:72]: "launch"

File: db/upgrade.php [hunk 12]
```diff
@@ -57,5 +54,36 @@ function xmldb_availability_maxviews_upgrade($oldversion) {
...
     }
+
+    if ($oldversion < 2023051717) {
+        // Define table availability_maxviews already created.
+        $table = new xmldb_table('availability_maxviews');
+        // Define the new field for overrider id.
+        $field = new xmldb_field('overriderid', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'userid');
+        $field->setComment('The id of the user who did the overriding process.');
+
+        // Conditionally lunch add field.
```
Comment (danielneis) [inline db/upgrade.php:65]: "launch".
Haven't you generated this code with the XMLDB Editor web interface?

File: delete.php [hunk 13]
```diff
@@ -23,23 +23,43 @@
  */
 
 require_once('../../../config.php');
+require_once(__DIR__.'/classes/event/maxviews_override_deleted.php');
```
Comment (danielneis) [inline delete.php:26]: as the namespace is correctly set on the class, it shouldn't be needed.
have you tried it without this line? (after cleaning the caches)

File: delete.php [hunk 14]
```diff
@@ -23,23 +23,43 @@
...
 
 $id = required_param('id', PARAM_INT);
 $courseid = required_param('courseid', PARAM_INT);
 $confirm = optional_param('confirm', 0, PARAM_BOOL);
 
 require_login();
-
+$context = context_course::instance($courseid);
 $ctx = context_system::instance();
```
Comment (danielneis) [inline delete.php:34]: you can delete $ctx, please

File: index.php [hunk 15]
```diff
@@ -27,13 +27,13 @@
 $courseid = required_param('courseid', PARAM_INT);
 
 require_login($courseid);
-
+$context = context_course::instance($courseid);
 $ctx = context_system::instance();
```
Comment (danielneis) [inline index.php:31]: you can delete $ctx, please

File: lang/en/availability_maxviews.php [hunk 16]
```diff
@@ -22,24 +22,47 @@
...
 $string['overridedeleted'] = 'Override deleted';
+$string['overridetime'] = 'Overrided at';
+$string['overridetype'] = 'Type of override';
+$string['overridetype_desc'] = 'Normal override: discarding the old condition and set a new one for the selected users.
+
+Adding maxviews: This mean that you add or subtract the new value from the old maxviews.';
 $string['overrides'] = 'Overrides';
 $string['overrideupdated'] = 'Override updated';
 $string['participant'] = 'Participant';
+$string['participant_help'] = 'Select the participants needed to override their maximum views for the selected course module.';
```
Comment (danielneis) [inline lang/en/availability_maxviews.php:60]: please change to "the participants you want to override the maximum views"

File: override.php [hunk 17]
```diff
@@ -23,14 +23,17 @@
  */
 
 require_once('../../../config.php');
+require_once(__DIR__.'/classes/event/maxviews_override_updated.php');
+require_once(__DIR__.'/classes/event/maxviews_override_created.php');
```
Comment (danielneis) [inline override.php:27]: these should't be needed as well as the namespace is correctly defined on both classes

File: override.php [hunk 18]
```diff
@@ -23,14 +23,17 @@
...
 
 $courseid = required_param('courseid', PARAM_INT);
 $id = optional_param('id', 0, PARAM_INT);
+$context = context_course::instance($courseid);
 
 require_login($courseid);
 
 $ctx = context_system::instance();
```
Comment (danielneis) [inline override.php:35]: plese, delete $ctx

File: override.php [hunk 19]
```diff
@@ -46,37 +49,104 @@
...
+            $eventarray['relateduserid'] = $userid;
+            // Prevent duplication of records for same user which makes a conflict at error when calling get_record() later.
+            if ($oldrecord = $DB->get_record('availability_maxviews', $record)) {
+                $newrecord = array_merge($newrecord,
+                                ['id' => $oldrecord->id,
+                                'timeupdated' => time(),
+                                ]);
+                $recordobject = (object)$newrecord;
+                // Adding the old maxviews.
+                $recordobject->maxviews += $oldrecord->maxviews;
```
Comment (danielneis) [inline override.php:124]: I think we should be testing for override type here, right?

File: override.php [hunk 20]
```diff
@@ -46,37 +49,104 @@
...
+            } else {
+                $newrecord['lastreset'] = 0;
+                $eventarray['other']['reset'] = false;
+            }
+            $eventarray['relateduserid'] = $userid;
+            // Prevent duplication of records for same user which makes a conflict at error when calling get_record() later.
+            if ($oldrecord = $DB->get_record('availability_maxviews', $record)) {
+                $newrecord = array_merge($newrecord,
+                                ['id' => $oldrecord->id,
+                                'timeupdated' => time(),
```
Comment (danielneis) [inline override.php:120]: please, don't use array_merge, here.
you can simply assign $newrecord->id = $oldrecord->id , and $newrecord->timeupdated = time()

File: override.php [hunk 21]
```diff
@@ -46,37 +49,104 @@
...
+                                ]);
+                $recordobject = (object)$newrecord;
```
Comment (danielneis) [inline override.php:122]: please, don't creat a new variable here, and cast only if really needed

File: override.php [hunk 22]
```diff
@@ -46,37 +49,104 @@
...
+                // Adding the old maxviews.
+                $recordobject->maxviews += $oldrecord->maxviews;
+                // Check if there is old reset first.
+                $recordobject->lastreset = max($recordobject->lastreset, $oldrecord->lastreset);
+                $DB->update_record('availability_maxviews', $recordobject);
```
Comment (danielneis) [inline override.php:127]: you can cast $newobject here when calling update_record

File: override.php [hunk 23]
```diff
@@ -46,37 +49,104 @@
...
+                $DB->insert_record('availability_maxviews', $recordobject);
+
+                $eventarray['other']['type'] = 'created';
+
+            }
+        }
         $msg = get_string('overrideadded', 'availability_maxviews');
     }
+    // Check the type of the event.
+    if ($eventarray['other']['type'] == 'updated') {
```
Comment (danielneis) [inline override.php:143]: you don't need an extra if, you can create the event on line after you define $eventarray[other][type]

File: templates/index.mustache [hunk 24]
```diff
@@ -43,6 +51,8 @@
               <td>{{userfullname}}</td>
               <td>{{maxviews}}</td>
               <td>{{#lastreset}}{{#userdate}} {{lastreset}} , {{#str}} strftimedatetime, core_langconfig {{/str}} {{/userdate}}{{/lastreset}}</td>
+              <td>{{modifier}}</td>
```
Comment (danielneis) [inline templates/index.mustache:54]: modifier or overrider?

File: classes/output/index.php [hunk 25]
```diff
@@ -60,11 +64,19 @@ public function export_for_template(renderer_base $output) {
             $overrides[$key]->deleteoverrideurl = $deleteoverrideurl->out(false);
             $overrides[$key]->coursemodule = $modinfo->cms[$o->cmid]->get_formatted_name();
             $overrides[$key]->userfullname = fullname($o);
+
+            // The user that did the override process.
+            $modifer = \core_user::get_user($o->overriderid);
```
Comment (danielneis) [inline classes/output/index.php:69]: please, use overrider

File: classes/output/index.php [hunk 26]
```diff
@@ -60,11 +64,19 @@ public function export_for_template(renderer_base $output) {
...
+            $overrides[$key]->modifier = fullname($modifer);
+            $overrides[$key]->date = max($o->timecreated, $o->timeupdated);
```
Comment (danielneis) [inline classes/output/index.php:71]: we can always use timeupdated, just make sure to set it on create time too.

**Review by danielneis — CHANGES_REQUESTED** [review]
Hello, Mo Farouk
Thank you very much for your contributions.
The good looks very good and consistent.
It only needs a few fixes and improvements to be ready to merge.
Please, take a look at the comments and if you have any questions, feel free to ask.
Best regards,
Daniel

File: classes/form/override.php (hunk 1 above)
Comment (fmido88) [reply classes/form/override.php:50]: Thanks, I'll use this instate:
```
            if (!empty($i->availability)) {
                $av = json_decode($i->availability);
                foreach ($av->c as $condition) {
                    if ($condition->type !== 'maxviews') {
                        continue;
                    }
                }
            } else {
                continue;
            }
```

File: lang/en/availability_maxviews.php (hunk 5 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php:63]: done

File: lang/en/availability_maxviews.php (hunk 4 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php:38]: done

File: lang/en/availability_maxviews.php (hunk 3 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php:37]: done

File: lang/en/availability_maxviews.php (hunk 2 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php:35]: done

File: lang/en/availability_maxviews.php (hunk 6 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php:64]: done

File: classes/form/override.php (hunk 8 above)
Comment (fmido88) [reply classes/form/override.php:121]: done

File: classes/output/index.php (hunk 9 above)
Comment (fmido88) [reply classes/output/index.php:79]: predefine the url

File: db/upgrade.php (hunk 10 above)
Comment (fmido88) [reply db/upgrade.php:79]: I think I was hungry :)

File: db/upgrade.php (hunk 12 above)
Comment (fmido88) [reply db/upgrade.php:65]: No I generated it manually

File: delete.php (hunk 13 above)
Comment (fmido88) [reply delete.php:26]: When I tried it for the first time, there was an error that the file not found, but you are right maybe I forgot to clean the cash.

File: delete.php (hunk 14 above)
Comment (fmido88) [reply delete.php:34]: done

File: index.php (hunk 15 above)
Comment (fmido88) [reply index.php:31]: We are using it in page context, but it's ok I'll change the page context.

File: lang/en/availability_maxviews.php (hunk 16 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php:60]: done

File: override.php (hunk 18 above)
Comment (fmido88) [reply override.php:35]: done

File: override.php (hunk 19 above)
Comment (fmido88) [reply override.php:124]: I didn't create a behat test before. I'll try to do.
but I tested it manually and it worked find

File: override.php (hunk 23 above)
Comment (fmido88) [reply override.php:143]: done

File: classes/output/index.php (hunk 26 above)
Comment (fmido88) [reply classes/output/index.php:71]: on the first override creation there is no timeupdated, it is equal to 0.
I'll modify it in the creation code, so timeupdated will set to time() at creation too

**Comment by fmido88** [comment]
@danielneis All done

File: classes/form/override.php [hunk 27]
```diff
@@ -40,33 +44,86 @@ public function definition() {
...
-            $options[$u->id] = fullname($u);
+
+        // Adding the user's identity fields to make it easier to serch for many users.
+        $context = \context_course::instance($this->_customdata['courseid']);
+        foreach ($users as $user) {
+            // Filter users with those with restrictions only.
+            if (has_capability('moodle/course:ignoreavailabilityrestrictions', $context, $user)) {
+                continue;
+            }
+            $userfields = \core_user\fields::get_identity_fields($context, true);
```
Comment (danielneis) [inline classes/form/override.php:85]: this could be outside of the foreach loop

File: classes/output/index.php (hunk 26 above)
Comment (danielneis) [reply classes/output/index.php:73]: did you remember to always update timeupdated ? we have to remove this max call here

**Review by danielneis — CHANGES_REQUESTED** [review]
thanks! most of thigs were accomplished. there's just a couple of little fixes to do.
I'll try the code locally next week. If you can write some behat tests would be awesome, but there's no problem if you don't.
Best,
Daniel

File: classes/output/index.php (hunk 26 above)
Comment (fmido88) [reply classes/output/index.php:73]: Yes, in override.php in any update action (but not in creation).

Comment (danielneis) [reply classes/output/index.php:73]: please, set his field in the creation process so we do not need this max here

File: classes/form/override.php (hunk 27 above)
Comment (fmido88) [reply classes/form/override.php:85]: You are right, I'll move it now.

File: classes/output/index.php (hunk 26 above)
Comment (fmido88) [reply classes/output/index.php]: Done

File: override.php [hunk 28]
```diff
@@ -39,44 +38,109 @@
...
+            $record = [
+                'courseid' => $data->courseid,
+                'cmid' => $data->cmid,
+                'userid' => $userid,
+            ];
+
+            $newrecord = [
+                'maxviews' => $data->maxviews,
+                'overriderid' => $USER->id,
+                'timeupdated' => time(),
```
Comment (danielneis) [inline override.php:104]: we don't have timecreated here too?

**Review by danielneis — CHANGES_REQUESTED** [review]
I think we're missing timecreated on the creation of the override. could you please double check and add if needed?

File: override.php (hunk 28 above)
Comment (fmido88) [reply override.php:104]: in line 130, because I use it in inserting records only.

**Comment by danielneis** [comment]
I've tried the code locally and experienced a few errors.
classes/output/index.php : line 71 should be like this:
            // The user that did the override process.
            if (empty($o->overriderid)) {
                $overrides[$key]->overrider = ''; 
                $overrides[$key]->date = get_string('never');
            } else {
                $overrider = \core_user::get_user($o->overriderid);
                $overrides[$key]->overrider = fullname($overrider);
                $overrides[$key]->date = $o->timeupdated;
            }   

Because for existing overrides there are no overriderid.

I also tried to edit an override that existed for the admin user (that is enrolled in course as student) and it didn't appear on the list. we should fetch the full list of participants with no extra filter.

Then I tried to edit another existing override but got an error :
Exception - Call to a member function trigger() on null
Output buffer:  <br /> <b>Warning</b>: Undefined variable $event in <b>moodle/availability/condition/maxviews/override.php</b> on line <b>143</b><br /> 

Please, fix this errors and re-submit.
If you can add any behat tests to make sure everything is keep working would be awesome.

**Comment by fmido88** [comment]
@danielneis sorry for the errors.
The first one I didn't put it in count and didn't appear to me as I worked with overriderid from the beginning so I missed that some used the plugin without this field. I fixed it.

The second problem about the filtration: I'm not convinced  about overriding a user who already can ignore restriction. Though I added an option for the filtration in the settings of the plugin. I hope this is ok.

The third problem: missing definition of $event. Fixed.

About the behat test, I'm really don't have a background about how to create a behat test.

**Comment by danielneis** [comment]
Hello,
About the multiple user select:
At first, I've found it usefull to create multiple overrides at once.
But then, I've found it brings more problems than it solves.
When editing an override, I should not be able to select the user I'm editing and, for sure, not selecting more than one.
When I select more than one, it looks like only the first is overriden.
When I create an override for multiple users, if there are already overrides for some, they are not updated, only the non existent one is created. It should update the existent or have a way to not allow to select users that already has overrides.
In this sense, you can think about group overrides, like assignment and quiz have, would be an option. 
If you can split your code and submit only the part of the events/logs, I can integrate that.
If you can split your code and submit only the part with the setting and the ability to sum the override, I can integrate that.
In summary, if you remove this part about multiple users I can integrate the rest.
Than you can elaborate more on this topic and resubmit.
Thank you for your collaboration and patience.
Best,
Daniel

**Comment by danielneis** [comment]
I've forgot to mention that on my last test, the "Overrided by" column was not displaying anything. The table has more headers than columns on the rest of the lines.

**Comment by fmido88** [comment]
Hello @danielneis 
I'm sorry for late response
I made some changes and I'll upload it soon regarding updating existence overrides.
For the "Overrided by" column I discovered a mistake in the mustache file and fixed it.

**Comment by fmido88** [comment]
In order to show in the list of users the only ones that hasn't been overridden yet, I think we must make a multiple pages form, The first one is to select the course modules and the second one for users and the rest of the form.

We can also add settings for either bulk override or just single user selection.

**Comment by fmido88** [comment]
Hello @danielneis I hope you are doing well.
Please check the latest updates.

File: lang/en/availability_maxviews.php (hunk 7 above)
Comment (fmido88) [reply lang/en/availability_maxviews.php]: Done
